artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -604,26 +607,39 @@ public String getAddress() {
       return addressSimpleString == null ? null : addressSimpleString.toString();
    }
 
+
+   public SimpleString cachedAddressSimpleString(String address) {
+      return CoreMessageObjectPools.cachedAddressSimpleString(address, coreMessageObjectPools);
+   }
+
    @Override
    public AMQPMessage setAddress(String address) {
-      this.address = SimpleString.toSimpleString(address, coreMessageObjectPools == null ? null : coreMessageObjectPools.getAddressStringSimpleStringPool());
+      setAddress(cachedAddressSimpleString(address));
       return this;
    }
 
    @Override
    public AMQPMessage setAddress(SimpleString address) {
       this.address = address;
+      createExtraProperties().putSimpleStringProperty(ADDRESS_PROPERTY, address);
       return this;
    }
 
    @Override
    public SimpleString getAddressSimpleString() {
       if (address == null) {
-         Properties properties = getProtonMessage().getProperties();
-         if (properties != null) {
-            setAddress(properties.getTo());
-         } else {
-            return null;
+
+         address = createExtraProperties().getSimpleStringProperty(ADDRESS_PROPERTY);
+
+         if (address != null) {
+            return address;
+         }
+
+
+         Properties properties = getProperties();
+         if (properties != null && properties.getTo() != null) {
+            address = cachedAddressSimpleString(properties.getTo());
+            return address;
```

Comment:
is this needed as anyhow last statement returns the address.

Reply:
I like having it explicitly as it shows my real intention.  Other than my ocd level not really needed.

Reply:
i personally prefer Single Entry, Single Exit, but as you note its all personal pref :), no worries if you meant to do it on purpose, just was checking it wasn't by mistake.

Reply:
I'd typically let it fall through here too, though in this case it really doesn't look that out of place considering the prior null check and exit above it already from checking the extra properties.

Reply:
@michaelandrepearce @gemmellr this made me realize a real issue though. I was always creating the extraProperties even when not needed. I added a better check now.. it's amended already.